# Movies Tonight Part 3: SQL DML

In [2]:
%load_ext sql
%sql sqlite:///Movies.db

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [3]:
%%sql
INSERT INTO MOVIES (MTitle, Rating)
    SELECT DISTINCT MTitle, Rating
    FROM DATASET ;

 * sqlite:///Movies.db
23 rows affected.


[]

In [4]:
%%sql
SELECT *
FROM MOVIES

 * sqlite:///Movies.db
Done.


[(1, 'Associate, The', 'PG-13'),
 (2, 'Ghost & The Darkness, The', 'R'),
 (3, 'Independence Day', 'PG-13'),
 (4, 'D3: The Mighty Ducks', 'PG'),
 (5, 'Dear God', ''),
 (6, 'First Wives Club, The', 'PG-13'),
 (7, 'High School High', 'PG-13'),
 (8, 'Larger Than Life', 'PG'),
 (9, 'Mirror Has Two Faces, The', 'PG-13'),
 (10, 'Ransom', 'R'),
 (11, 'Sleepers', ''),
 (12, 'Space Jam', 'PG'),
 (13, "Stephen King's Thinner", ''),
 (14, 'Michael Collins', 'R'),
 (15, 'Romeo & Juliet', 'R'),
 (16, 'Big Night', 'R'),
 (17, 'Fly Away Home', 'PG'),
 (18, 'Nutty Professor, The', 'PG-13'),
 (19, 'Phenomenon', 'PG'),
 (20, 'Set It Off', 'R'),
 (21, 'Surviving Picasso', 'R'),
 (22, 'Very Brady Sequel, A', 'PG-13'),
 (23, 'That Thing You Do', 'PG')]

In [5]:
%%sql 
INSERT INTO THEATER (TName, Phone, Location)
    SELECT DISTINCT TName, Phone, Location
    FROM DATASET ;

 * sqlite:///Movies.db
9 rows affected.


[]

In [6]:
%%sql
SELECT *
FROM THEATER

 * sqlite:///Movies.db
Done.


[(1, 'Akarakian Theatres Moreno 4 Cinemas', '(909) 485-2899', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley'),
 (2, 'Cinema Star The Ultraplex 14', '(909) 342-2256', 'Mission Grove, Riverside'),
 (3, 'General Cinema Rancho 6', '(714) 370-2085', ' I-215 At Mt. Vernon S. At I-10, San Bernardino'),
 (4, 'Pacific Inland Center', '(714) 381-1611', ' Inland Center Mall, San Bernardino'),
 (5, 'SOCAL Canyon Crest Cinema', '(909) 682-6900', 'Central Avenue South Of 60 Freeway Near Ucr, Riverside'),
 (6, 'SOCAL Canyon Springs Cinema', '(909) 782-0800', 'East Of I-215 On 60 Freeway At Day Street Canyon, Moreno Valley'),
 (7, 'SOCAL Marketplace Cinema', '(909) 682-4040', 'University/mission Inn Exits East Of 91 Freeway On, Riverside'),
 (8, 'United Artists Riverside (Galleria) Tyler Mall', '(714) 689-802', 'Riverside Fwy Tyler, Riverside'),
 (9, 'United Artists Riverside Park Sierra', '(909) 359-6995', '3600 Park Sierra Dr., Riverside')]

In [7]:
%%sql
INSERT INTO ARTIST (AName)
SELECT DISTINCT CName 
    FROM DATASET ;

 * sqlite:///Movies.db
153 rows affected.


[]

In [9]:
%%sql 
INSERT INTO CREDIT (MID, AID, CCode)
SELECT DISTINCT MID, CCode, AID 
    FROM DATASET
    JOIN Movies ON (Movies.MTitle=DATASET.MTitle)
    JOIN Artist ON (Artist.AName=DATASET.CName) ;

 * sqlite:///Movies.db
162 rows affected.


[]

In [12]:
%%sql
INSERT INTO SHOW (MID, TID, STime)
SELECT DISTINCT MID, TID, ShowTime
FROM DATASET
    JOIN Movies ON (DATASET.MTitle=Movies.MTitle)
    JOIN Theater ON (DATASET.TName=Theater.TName) ;

 * sqlite:///Movies.db
131 rows affected.


[]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

In [15]:
%%sql
SELECT COUNT (*)
    FROM MOVIES;

 * sqlite:///Movies.db
Done.


[(23,)]

In [16]:
%%sql    
SELECT COUNT(*)
    FROM Theater;

 * sqlite:///Movies.db
Done.


[(9,)]

In [17]:
%%sql    
SELECT COUNT(*)
    FROM Artist;
 

 * sqlite:///Movies.db
Done.


[(153,)]

In [18]:
%%sql  
SELECT COUNT(*)
    FROM Show;
 

 * sqlite:///Movies.db
Done.


[(131,)]

In [19]:
 %%sql  
SELECT COUNT(*)
    FROM Credit;

 * sqlite:///Movies.db
Done.


[(162,)]

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [25]:
%%sql 
SELECT NONDISTINCT COUNT(*)
    FROM DATASET;

 * sqlite:///Movies.db
(sqlite3.OperationalError) near "(": syntax error [SQL: 'SELECT NONDISTINCT COUNT(*)\n    FROM DATASET;'] (Background on this error at: http://sqlalche.me/e/e3q8)


In [26]:
%%sql 
SELECT COUNT(*)
    FROM DATASET ;

 * sqlite:///Movies.db
Done.


[(913,)]

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".